tools/fork-network/src/trim_database.rs

```diff
@@ -0,0 +1,159 @@
+use indicatif::HumanBytes;
+use near_chain::{ChainStore, ChainStoreAccess};
+use near_chain_configs::GenesisConfig;
+use near_epoch_manager::{EpochManager, EpochManagerAdapter};
+use near_primitives::{borsh::BorshDeserialize, state::FlatStateValue};
+use near_store::flat::store_helper::{decode_flat_state_db_key, encode_flat_state_db_key};
+use near_store::parallel_iter::StoreParallelIterator;
+use near_store::{DBCol, Store};
+
+/// Trims the database for forknet by rebuilding the entire database with only the FlatState
+/// and the small part of State that is needed for flat state, i.e. the non-inlined trie values.
+/// Also copies over the basic DbVersion and Misc columns.
+///
+/// Does not mutate the old store, but populates the new store with the trimmed data.
+///
+/// Note that only the state in the FlatState is copied. All other state in FlatStateChanges is
+/// ignored. For this reason, this is only suitable for forknet testing.
+pub fn trim_database(old: Store, genesis_config: &GenesisConfig, new: Store) -> anyhow::Result<()> {
+    // We'll only copy over the data corresponding to shard UIDs for the latest epoch.
+    let sharding_version = {
+        let epoch_manager = EpochManager::new_arc_handle(old.clone(), genesis_config);
+        let chain = ChainStore::new(old.clone(), genesis_config.genesis_height, false);
+        let final_head = chain.final_head()?;
+        let epoch_id = final_head.epoch_id;
+        let shard_layout = epoch_manager.get_shard_layout(&epoch_id)?;
+        shard_layout.version()
+    };
+
+    // We'll parallel-iterate over the flat storage and also write to the new store at the same time.
+    // I'm not sure if there is value in parallelizing the write side as well, as that is now the
+    // bottleneck.
+    let (tx, rx) = std::sync::mpsc::sync_channel(1000000);
+    let iter_handle = {
+        let old = old.clone();
+        std::thread::spawn(move || {
```

Comment:
not a big deal, but why do we need a new thread for this? this new thread will just be spawning other ones and then waiting in join() right? again, not a big deal, but could just do the joining later when you feel like it right?

Reply:
Well both sides of the channel needs to be operated on in parallel. So either this is in a thread, or the consumer side is in a thread.